### PR TITLE
`yarn` → `pnpm`

### PR DESCRIPTION
**Pull request by sndrs**
### Swaps `yarn` for `pnpm`

`pnpm` is much faster to install, much lighter on disk, and because of that it is able isolate dependencies so that no transitive deps are available. 

Updating the apps to get this working has surfaced real bugs and instability. 

Using `pnpm` should hopefully prevent this class of bug recurring, while keeping a useable, predictable `node_modules` dirs around (unlike, for example, [yarn's pnp mode](https://yarnpkg.com/features/pnp), which solves the problem in a more non-standard way).

> [!NOTE]
> To get the build working, I had to create the `pnpm-lock.yaml` from scratch – importing the current `yarn.lock` resulted in various issues with webpack types (and i think others too) which starting clean resolved.
>
> This means that dependencies will have been bumped to the latest available versions in the specified ranges. This is probably a good thing, but it does mean that build output is not going to be identical to `main`. I _think_ that's ok?

| Installing `node_modules` | Before | After |
| ----------- | ----------- | ---------- |
| `make install` (no changes) | 2.9s | 1.7s |
| `./.github/actions/setup-node-env` (with a cache hit) | 51s | 20s |

### Removes the `preinstall` corepack check

Only yarn runs the `preinstall` hook _before_ installing (see [here](https://github.com/pnpm/pnpm/issues/3836) and [here](https://github.com/npm/cli/issues/2660)), so there's not a lot of point in keeping this around.

### CODE REVIEW

File: dotcom-rendering/src/components/SignInGate/README.md [hunk 1]
```diff
@@ -267,7 +267,7 @@ mainPatientia.storyName = 'patientia_standalone';
 
 The props can be set to anything here, since storybook is only interested in the design of the component.
 
-To view it in storybook simply run `yarn storybook` which will launch a storybook server (with live reloading) to be able to view and develop it.
+To view it in storybook simply run `pnpm storybook` which will launch a storybook server (with live reloading) to be able to view and develop it.
```
Comment (mxdvl) [inline dotcom-rendering/src/components/SignInGate/README.md:270]:
```suggestion
To view it in storybook simply run `make storybook-dev` which will launch a storybook server (with live reloading) to be able to view and develop it.
```

File: scripts/deno/peer-dependencies.ts
Comment (mxdvl) [inline scripts/deno/peer-dependencies.ts:1]: I think this file should go!

File: dotcom-rendering/makefile [hunk 2]
```diff
@@ -104,22 +104,22 @@ tsc: clean-dist install
...
-	@cd .. && yarn prettier:write
+	@cd .. && pnpm prettier:write
 
 fix: clear clean-dist install lint-fix prettier
 	$(call log, "everything fixed")
 
 snapshot: clear clean-dist install
 	$(call log, "taking snapshots")
-	yarn chromatic
+	pnpm chromatic
```
Comment (mxdvl) [inline dotcom-rendering/makefile:118]:
```suggestion
	@pnpm chromatic
```

File: package.json [hunk 3]
```diff
@@ -28,23 +21,24 @@
 	},
 	"lint-staged": {
 		"dotcom-rendering/**/*.md": [
-			"yarn workspace @guardian/dotcom-rendering createtoc",
-			"prettier --write --cache"
+			"pnpm --filter @guardian/dotcom-rendering createtoc",
+			"prettier --write"
 		],
-		"*": "prettier --ignore-unknown --write --cache"
+		"*": "prettier --ignore-unknown --write"
```
Comment (mxdvl) [inline package.json:27]: Why is the `--cache` gone?

File: apps-rendering/package.json [hunk 4]
```diff
@@ -8,19 +8,18 @@
...
-		"test:coverage": "yarn test; open coverage/index.html",
+		"test:coverage": "pnpm test; open coverage/index.html",
 		"test:watch": "jest --config config/jest.config.js --watch",
 		"lint": "eslint -c config/.eslintrc.js ./ --ext '.ts,.tsx' --cache",
 		"lint:fix": "eslint -c config/.eslintrc.js ./ --ext '.ts,.tsx' --fix --cache",
 		"watch:server": "webpack --config webpack.config.ts --config-name server --env watch",
 		"watch:client": "webpack serve --config webpack.config.ts --config-name client",
 		"watch:cdk": "tsc -w -p config/tsconfig.cdk.json",
-		"watch": "yarn watch:server & yarn watch:client",
+		"watch": "pnpm '/^watch:(?!cdk$).*/'",
```
Comment (mxdvl) [inline apps-rendering/package.json:22]: Closer to the original:

```suggestion
		"watch": "pnpm '/^watch:(client|server)/'",
```

**Review by mxdvl — APPROVED** [review]
Let’s make this package management performant!

(left three teeny-tiny comments, which can be addressed in a follow-up)

File: apps-rendering/package.json (hunk 4 above)
Comment (sndrs) [reply apps-rendering/package.json:22]: yeah much nicer

File: package.json (hunk 3 above)
Comment (sndrs) [reply package.json:27]: i'm not what it offers here - we know the file has changed, so i don't think we will never want to skip formatting in this instance?

Comment (sndrs) [reply package.json:27]: though admittedly it's not pnpm-related and could be moved to a separate change

**Comment by sndrs** [comment]
@mxdvl sorry, added one more change in f1105f4

**Review by arelra — APPROVED** [review]
🚀 Very nice performance bump and the culmination of lots of prior work.

Just as a precaution could we deploy this to CODE pls to make sure the prod build and start is not affected. I will raise a ticket to add a GHA check to run the prod task so this is automated in future.

**Comment by sndrs** [comment]
> Just as a precaution could we deploy this to CODE pls to make sure the prod build and start is not affected. I will raise a ticket to add a GHA check to run the prod task so this is automated in future.

DCR deployed to CODE here https://riffraff.gutools.co.uk/deployment/view/0d3751e5-0a93-4d50-8961-5fb5c7cd0ec3

checking AR...

File: dotcom-rendering/package.json [hunk 5]
```diff
@@ -5,26 +5,25 @@
 	"license": "Apache-2.0",
 	"private": true,
 	"scripts": {
-		"preinstall": "../scripts/preinstall.mjs",
-		"lint": "yarn lint:src && yarn lint:cypress",
-		"lint:fix": "yarn lint:src --fix && yarn lint:cypress --fix",
+		"lint": "pnpm '/^lint:(cypress|scripts|src)$/'",
+		"lint:fix": "pnpm lint --fix",
```
Comment (mxdvl) [inline dotcom-rendering/package.json:9]: I don’t think this pases the arguments as expected

```
. lint:src$ eslint . --ext .ts,.tsx --quiet --cache "--fix"
└─ Running...
. lint:scripts$ eslint scripts --quiet --cache "--fix"
└─ Running...
. lint:cypress$ eslint ./cypress --cache "--fix"
└─ Done in 472ms
```

Comment (sndrs) [reply dotcom-rendering/package.json:9]: this shouldn't affect how it works - iiuc, all options are strings that are then coerced (or not) into types by the whatever code constitutes the program you're running (in this case eslint).

i think the issue predates pnpm, in that the eslint command `ext` flags have not updates include all files in the application (`.mjs` etc – by default eslint only looks for `.js`).

i'm also not 100% sure why there's a rule for everything (`.`) and 2 for directories that will also be included in that rule (`cypress` and `scripts`)?

Comment (mxdvl) [reply dotcom-rendering/package.json:9]: Yeah all fair points, probably one to address separately!

**Pull request merged**